README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# askgpt <img src="man/figures/logo.png" align="right" height="139" />

<!-- badges: start -->

[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-green.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable)
[![R-CMD-check](https://github.com/JBGruber/askgpt/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/JBGruber/askgpt/actions/workflows/R-CMD-check.yaml)
[![Codecov test
coverage](https://codecov.io/gh/JBGruber/askgpt/branch/main/graph/badge.svg)](https://app.codecov.io/gh/JBGruber/askgpt?branch=main)
[![CRAN
status](https://www.r-pkg.org/badges/version/askgpt)](https://CRAN.R-project.org/package=askgpt)
[![CRAN_Download_Badge](https://cranlogs.r-pkg.org/badges/grand-total/askgpt)](https://cran.r-project.org/package=askgpt)
<!-- badges: end -->

You’re new to R? You don’t quite understand the code you copied from
that tutorial? You get error messages that make no sense to you? Don’t
worry, just *askgpt*!

`askgpt` is basically ChatGPT but from R (technically, it sends prompts
to [OpenAI’s API](https://openai.com/api/) directly from R). It also has
some additional functionality:

- Prompt *“What is wrong with my last command?”* (or *“help!”*) to get
  help on the last error R emitted
- Use the RStudio addin to comment, annotate or explain highlighted code

See more under [Usage](#usage).

## Installation

You can install release version of askgpt like so:

``` r
install.packages("askgpt")
```

You can install the development version of askgpt like so:

``` r
remotes::install_github("JBGruber/askgpt")
```

## Log In

You can log into OpenAI’s API running the dedicated `login()` function
or just ask something with `askgpt()`:

``` r
library(askgpt)
login()
#> ℹ It looks like you have not provided an API key yet. Let me guide you through the process:
#>   1. Go to <https://platform.openai.com/account/api-keys>
#>   2. (Log into your account if you haven't done so yet)
#>   3. On the site, click the button + Create new secret key
#>   to create an API key
#>   4. Copy this key into R/RStudio
```

<div class="figure">

<img src="vignettes/figs/key.png" alt="Copy the API key from OpenAI's website" width="100%" />
<p class="caption">
Copy the API key from OpenAI’s website
</p>

</div>

<div class="figure">

<img src="vignettes/figs/key-rs.png" alt="And paste it into RStudio" width="100%" />
<p class="caption">
And paste it into RStudio
</p>

</div>

You will not need to do this again after the first time. (Technical
detail: This will store an encrypted version of your key in the
directory returned by `rappdirs::user_cache_dir("askgpt")`.) If your old
API key does not work any more, you can store a new one with:
`login(force_refresh = TRUE)`.

## Usage

To enable error logging (which you need if you want askgpt to explain
errors to you) first run:

``` r
log_init()
```

After this, the key phrase **“What is wrong with my last command?”** (or
just “help!”) will make `askgpt` look up your last command and error
message and return some help for you. The other important key phrase is
**“Can you elaborate on that?”** (or just “What?”), which will elaborate
on the previous answer. You can ask basically any question you want
though:

<figure>
<video
src="https://user-images.githubusercontent.com/23524101/220632749-da59fe90-d6a2-4ebd-bbfe-d1164e8a7944.mp4"
controls=""><a
href="https://user-images.githubusercontent.com/23524101/220632749-da59fe90-d6a2-4ebd-bbfe-d1164e8a7944.mp4">some
examples</a></video>
<figcaption aria-hidden="true">some examples</figcaption>
</figure>

## Configure

You can configure how `askgpt` makes that start with `askgpt_*`. For
example, to use a different model to use in `askgpt()` use
`options(askgpt_chat_model = "gpt-3.5-turbo-0301")` (or
`options(askgpt_chat_model = "gpt-4")` if you have beta access to the
newer model). If you use the completions instead of the chat API
(`chat = FALSE` in `askgpt()`) use
`options(askgpt_completions_model = "text-curie-001")`. It does not
matter if the API parameter is listed in the function or not. All are
used. See the complete list
[here](https://platform.openai.com/docs/api-reference/chat) and
[here](https://platform.openai.com/docs/api-reference/completions).

The most important setting, however, is `askgpt_config`. This can be
used to configure the chat using plain English:

``` r
options(askgpt_config = "I'm 8 years old, please explain things easily")
askgpt("What is an R function?")
#> 
#> ── Answer ──────────────────────────────────────────────────────────────────────
#> An R function is like giving your friend a set of instructions to perform a
#> particular task. In R programming, a function is a set of instructions or steps
#> that is given a name, and when you call that name, the function will perform
#> those instructions. A function can take information or inputs, do something
#> with those inputs (like adding or subtracting), and then give the result back
#> as output.
#> 
#> For example, think about giving your friend the instructions to make a peanut
#> butter sandwich. The instructions might be:
#> 
#> 1. Take two slices of bread 2. Spread peanut butter on one slice 3. Spread
#> jelly on the other slice 4. Put the two slices together
#> 
#> In R, a function might take a number (like 5) and add 1 to it, and then return
#> the result (which would be 6).
#> 
#> Functions in R are used to make code easier to use, understand, and reuse. They
#> can also help programmers write complex and efficient programs.
```


NEWS.md
# askgpt 0.1.3

- fixes bug that ignored some options when askgpt() was run with spinner
- adds an addin that can improve code, documentation, or writing given a prompt

# askgpt 0.1.2

- fixed bug that would keep the package from printing answers to screen 

# askgpt 0.1.1

- better error handling
- answers look better on screen and can handle Markdown code without breaking
- new_conversation() function introduces the ability to make GPT forget previous questions
- removes the streaming option, since it made little sense (see https://github.com/JBGruber/askgpt/issues/4)
- improved the tutorialise adding to handle long documents and gave it a nicer interface
- various bug fixes

# askgpt 0.0.2

First release to CRAN! This version introduces the core functionality of
askgpt, which goal it is to make R more approachable by employing the power of
the ChatGPT language model through the OpenAI APIs. The package provides the
main function askgpt(), which can be used as to ask questions to OpenAI's GPT
models, and some tools to document code via an RStudio Addin.


R/api.r
#' Request answer from openai's completions API
#'
#' Mostly used under the hood for \code{\link{askgpt}}.
#'
#' @param prompt character string of the prompt to be completed.
#' @param model character string of the model to be used (defaults to
#'   "text-davinci-003").
#' @param temperature numeric value between 0 and 1 to control the randomness of
#'   the output (defaults to 0.2; lower values like 0.2 will make answers more
#'   focused and deterministic).
#' @param max_tokens The maximum number of tokens to generate in the completion.
#'   2048L is the maximum the models accept.
#' @param api_key set the API key. If NULL, looks for the env OPENAI_API_KEY.
#' @param ... additional parameters to be passed to the API (see [the API
#'   documentation](https://platform.openai.com/docs/api-reference/completions)
#'
#' @details Only a few parameters are implemented by name. Most can be sent
#'   through the \code{...}. For example, you could use the \code{n} parameter
#'   just like this \code{completions_api("The quick brown fox", n = 2)}.
#'
#'   A couple of defaults are used by the package:
#'   \itemize{
#'      \item{the model used by default is "text-davinci-003"}
#'      \item{the default temperature is 0.2}
#'      \item{the default for max_tokens is 2048L}
#'   }
#'
#'   You can configure how \code{\link{askgpt}} makes requests by setting
#'   options that start with \code{askgpt_*}. For example, to use a different
#'   model use \code{options(askgpt_model = "text-curie-001")}. It does not
#'   matter if the API parameter ist listed in the function or not. All are
#'   used.
#'
#' @importFrom rlang `%||%`
#'
#' @return a httr2 response object
#' @export
#'
#' @examples
#' \dontrun{
#' completions_api("The quick brown fox")
#' }
completions_api <- function(prompt,
                            model = NULL,
                            temperature = NULL,
                            max_tokens = NULL,
                            api_key = NULL,
                            ...) {

  model <- model %||% getOption("askgpt_completions_model") %||% "text-davinci-003"
  temperature <- temperature %||% getOption("askgpt_temperature") %||% 0.2
  max_tokens <- max_tokens %||% getOption("askgpt_max_tokens") %||% 2048L
  api_key <- api_key %||% login()

  # collect additional options
  params <- list(...)

  if (!is.null(params$stream)) cli::cli_warn("The streaming feature has been removed from the package.")
  # Todo: find more elegant way to remove these
  params$hist <- NULL
  params$config <- NULL

  askopts <- grep("^askgpt_", names(.Options), value = TRUE) |>
    setdiff(c("askgpt_chat_model", "askgpt_completions_model", "askgpt_config",
              "askgpt_key", "askgpt_max_tokens", "askgpt_stream", "askgpt_temperature"))
  for (par in askopts) {
    params[gsub("askgpt_", "", par, fixed = TRUE)] <- getOption(par)
  }

  body <- c(list(
    model = model,
    prompt = prompt,
    temperature = temperature,
    max_tokens = max_tokens
  ), params)

  body <- Filter(Negate(is.null), body)

  req <- httr2::request("https://api.openai.com/v1/completions") |>
    httr2::req_method("POST") |>
    httr2::req_headers(
      "Content-Type" = "application/json",
      "Authorization" = glue::glue("Bearer {api_key}")
    ) |>
    httr2::req_body_json(body)

  resp <- httr2::req_perform(req) |>
      httr2::resp_body_json()

  resp$call <- req
  resp$api <- "completions"
  return(resp)
}


#' Request answer from openai's chat API
#'
#' @param config a configuration prompt to tell the model how it should behave.
#'
#' @inheritParams completions_api
#'
#' @return A tibble with available models
#'
#' @importFrom rlang `%||%`
#'
#' @return a httr2 response object
#' @export
#'
#' @examples
#' \dontrun{
#' chat_api("Hi, how are you?", config = "answer as a friendly chat bot")
#' }
chat_api <- function(prompt,
                     model = NULL,
                     config = NULL,
                     max_tokens = NULL,
                     api_key = NULL,
                     ...) {

  model <- model %||% getOption("askgpt_chat_model") %||% "gpt-3.5-turbo"
  config <- config %||% getOption("askgpt_config")
  max_tokens <- max_tokens %||% getOption("askgpt_max_tokens") %||% 2048L
  api_key <- api_key %||% login()

  # collect additional options
  params <- list(...)
  if (!is.null(params$stream)) cli::cli_warn("The streaming feature has been removed from the package.")
  askopts <- grep("^askgpt_", names(.Options), value = TRUE) |>
    setdiff(c("askgpt_chat_model", "askgpt_completions_model", "askgpt_config",
              "askgpt_key", "askgpt_max_tokens", "askgpt_stream", "askgpt_temperature"))
  for (par in askopts) {
    params[gsub("askgpt_", "", par, fixed = TRUE)] <- getOption(par)
  }

  hist <- params$hist %||% c(rbind(prompt_history(), response_history()))
  params$hist <- NULL

  messages <- dplyr::bind_rows(list(
    if (!is.null(config)) data.frame(role = "system",
                                     content = config),
    if (length(hist) > 0) data.frame(role = c("user", "assistant"),
                                     content = hist),
    if (!methods::is(prompt, "data.frame")) data.frame(role = "user",
                                                       content = prompt) else prompt
  ))

  body <- c(list(
    model = model,
    messages = messages,
    max_tokens = max_tokens
  ), params)

  body <- Filter(Negate(is.null), body)

  req <- httr2::request("https://api.openai.com/v1/chat/completions") |>
    httr2::req_method("POST") |>
    httr2::req_headers(
      "Content-Type" = "application/json",
      "Authorization" = glue::glue("Bearer {api_key}")
    ) |>
    httr2::req_body_json(body) |>
    httr2::req_error(body = error_body)

  resp <- httr2::req_perform(req) |>
    httr2::resp_body_json()
  resp$call <- req
  resp$api <- "chat"
  return(resp)
}

error_body <- function(resp) {
  msg <- httr2::resp_body_json(resp)$error$message
  if (httr2::resp_status(resp) == 401L) {
    msg <- c(msg, "This error can also mean that you ran out of credit. Check https://platform.openai.com/account/usage")
  }
  msg
}


#' List Models
#'
#' List the models available in the API. You can refer to the [Models
#' documentation](https://platform.openai.com/docs/models) to understand what
#' models are available and the differences between them.
#'
#'
#' @inheritParams completions_api
#'
#' @return A tibble with available models
#'
#' @importFrom rlang `%||%`
#'
#' @export
#'
#' @examples
#' \dontrun{
#' completions_api("The quick brown fox")
#' }
list_models <- function(api_key = NULL) {

  api_key <- api_key %||% login()

  req <- httr2::request("https://api.openai.com/v1/models") |>
    httr2::req_method("GET") |>
    httr2::req_headers(
      "Content-Type" = "application/json",
      "Authorization" = glue::glue("Bearer {api_key}")
    )

  resp <- httr2::req_perform(req) |>
    httr2::resp_body_json()

  dplyr::bind_rows(resp$data)
}
